cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java

```diff
@@ -61,20 +67,32 @@
 
   /**
    * Executes given {@link Program} with the given {@link ProgramOptions} and block until it completed.
-   * On completion, return the {@link RuntimeContext} of the program.
+   * On completion, return the {@link Map} of hadoop counters for the program run.
    *
    * @throws Exception if execution failed.
    */
   @Override
-  public RuntimeContext runAndWait(Program program, ProgramOptions options) throws Exception {
+  public Map<String, String> runAndWait(Program program, ProgramOptions options) throws Exception {
     ProgramController controller = programRunnerFactory.create(ProgramRunnerFactory.Type.MAPREDUCE).run(program,
                                                                                                         options);
     if (controller instanceof MapReduceProgramController) {
-      final RuntimeContext context = ((MapReduceProgramController) controller).getContext();
-      return executeProgram(controller, context);
+      MapReduceContext context = ((MapReduceProgramController) controller).getContext();
+      executeProgram(controller, context);
+      return getCounterMap(context);
     } else {
       throw new IllegalStateException("Failed to run program. The controller is not an instance of " +
                                         "MapReduceProgramController");
     }
   }
+
+  private Map<String, String> getCounterMap(MapReduceContext context) throws Exception {
+    Map<String, String> mapCounters = Maps.newHashMap();
+    Counters counters = ((Job) context.getHadoopJob()).getCounters();
+    for (CounterGroup group : counters) {
+      for (Counter counter : group) {
+        mapCounters.put(group.getName() + "." + counter.getName(), Long.toString(counter.getValue()));
```

Comment:
We are just concatenating the counter group name and counter name to represent the key in the map. We are not recreating group name and counter name from the key, so it does not matter even if the counter name has `.` in it.
e.g. assume the group name is `some.group.name` and counter name is `some.counter.name`. The key in the map will be `some.group.name.some.counter.name`.  We will document the same.